gateway/sds_gateway/visualizations/processing/waterfall.py

```diff
@@ -0,0 +1,266 @@
+"""Waterfall processing logic for visualizations."""
+
+import base64
+import datetime
+import os
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any
+
+import h5py
+import numpy as np
+from digital_rf import DigitalRFReader
+from django.conf import settings
+from loguru import logger
+
+from sds_gateway.api_methods.utils.minio_client import get_minio_client
+
+
+@dataclass
+class WaterfallSliceParams:
+    """Parameters for processing a waterfall slice."""
+
+    reader: Any
+    channel: str
+    slice_idx: int
+    start_sample: int
+    samples_per_slice: int
+    end_sample: int
+    fft_size: int
+    min_frequency: float
+    max_frequency: float
+    sample_rate: float
+    center_freq: float
+
+
+def _process_waterfall_slice(params: WaterfallSliceParams) -> dict[str, Any] | None:
+    """Process a single waterfall slice."""
+    # Calculate sample range for this slice
+    slice_start_sample = (
+        params.start_sample + params.slice_idx * params.samples_per_slice
+    )
+    slice_num_samples = min(
+        params.samples_per_slice, params.end_sample - slice_start_sample
+    )
+
+    if slice_num_samples <= 0:
+        return None
+
+    # Read the data
+    data_array = params.reader.read_vector(
+        slice_start_sample, slice_num_samples, params.channel, 0
+    )
+
+    # Perform FFT processing
+    fft_data = np.fft.fft(data_array, n=params.fft_size)
+    power_spectrum = np.abs(fft_data) ** 2
+
+    # Convert to dB
+    power_spectrum_db = 10 * np.log10(power_spectrum + 1e-12)
+
+    # Convert power spectrum to binary string for transmission
+    data_bytes = power_spectrum_db.astype(np.float32).tobytes()
+    data_string = base64.b64encode(data_bytes).decode("utf-8")
+
+    # Create timestamp from sample index
+    timestamp = datetime.datetime.fromtimestamp(
+        slice_start_sample / params.sample_rate, tz=datetime.UTC
+    ).isoformat()
```

Comment:
maybe most of these could be computed_properties of `params`, if it was a pydantic model